python/ray/serve/scripts.py

```diff
@@ -232,49 +234,170 @@ def start(
+    default=None,
+    required=False,
+    help=(
+        "Directory containing files that your application(s) will run in. When using "
+        "the 'local' provider, this must be a remote URI to a .zip file (e.g., S3 "
+        "bucket). Other providers may support local paths. This overrides the "
+        "working_dir in --runtime-env if both are specified."
+    ),
+)
+@click.option(
+    "--name",
+    required=False,
+    default=None,
+    type=str,
+    help="Custom name for the application(s).",
+)
+@click.option(
+    "--base-image",
+    required=False,
+    default=None,
+    type=str,
+    help=(
+        "Container image to use for the application(s). Not supported for the "
+        "default 'local' provider because it deploys to an existing cluster."
+    ),
+)
+@click.option(
+    "--provider",
+    required=False,
+    default=None,
+    type=str,
+    help=(
+        "Deploy provider to use. By default, this uses a 'local' provider that makes "
+        "a REST API request to the provided address. Can also be set with the "
+        f"{DEPLOY_PROVIDER_ENV_VAR} environment variable."
     ),
 )
-@click.argument("config_file_name")
 @click.option(
     "--address",
     "-a",
     default=os.environ.get("RAY_DASHBOARD_ADDRESS", "http://localhost:8265"),
     required=False,
     type=str,
-    help=RAY_DASHBOARD_ADDRESS_HELP_STR,
+    help=RAY_DASHBOARD_ADDRESS_HELP_STR + " . Only used by the 'local' provider.",
 )
-def deploy(config_file_name: str, address: str):
-    warn_if_agent_address_set()
-
-    with open(config_file_name, "r") as config_file:
-        config = yaml.safe_load(config_file)
+def deploy(
+    config_or_import_path: str,
+    arguments: Tuple[str],
+    runtime_env: str,
+    runtime_env_json: str,
+    working_dir: str,
+    name: Optional[str],
+    base_image: Optional[str],
+    provider: Optional[str],
+    address: str,
+):
+    args_dict = convert_args_to_dict(arguments)
+    final_runtime_env = parse_runtime_env_args(
+        runtime_env=runtime_env,
+        runtime_env_json=runtime_env_json,
+        working_dir=working_dir,
+    )
 
-    ServeDeploySchema.parse_obj(config)
-    ServeSubmissionClient(address).deploy_applications(config)
+    # Skip validating runtime_env URIs so publishers can enable local URI usage.
+    with _skip_validating_runtime_env_uris():
+        config = _generate_config_from_file_or_import_path(
+            config_or_import_path,
+            name=name,
+            arguments=args_dict,
+            runtime_env=final_runtime_env,
+        )
 
-    cli_logger.success(
-        "\nSent deploy request successfully.\n "
-        "* Use `serve status` to check applications' statuses.\n "
-        "* Use `serve config` to see the current application config(s).\n"
+    deploy_provider = get_deploy_provider(provider)
+    deploy_provider.deploy(
+        config.dict(exclude_unset=True),
+        address=address,
+        name=name,
+        ray_version=ray.__version__,
```

Comment:
Do we raise a nice exception if the local ray version doesn't match the ray version in the base image or the remote cluster?

Reply:
That's up to the provider. I think this version is actually totally unnecessary because the provider can just check `ray.__version__` themselves

Reply:
I deleted it